### PR TITLE
Add metrics to udp input

### PR DESCRIPTION
**Pull request by purbon**
Add metrics to track:
- # events processed.
- # workers created.
- # errors that happened during execution time (different types).

related to https://github.com/elastic/logstash/issues/5607 https://github.com/elastic/logstash/issues/5717

### CODE REVIEW

File: lib/logstash/inputs/udp.rb [hunk 1]
```diff
@@ -97,10 +101,12 @@ def inputworker(number)
           decorate(event)
           event.set("host", client[3]) if event.get("host").nil?
           @output_queue.push(event)
+          metric.increment(:events)
```
Comment (ph) [inline lib/logstash/inputs/udp.rb]: I think this should be done in the codec layer.

File: lib/logstash/inputs/udp.rb [hunk 2]
```diff
@@ -67,6 +69,8 @@ def udp_listener(output_queue)
     @udp.bind(@host, @port)
 
     @input_to_worker = SizedQueue.new(@queue_size)
+    metric.gauge(:queue_size, @queue_size)
+    metric.gauge(:threads, @workers)
```
Comment (ph) [inline lib/logstash/inputs/udp.rb]: I think `workers` would be a better name than `threads`, it reflects the name of the setting.

File: lib/logstash/inputs/udp.rb [hunk 3]
```diff
@@ -50,6 +51,7 @@ def run(output_queue)
       udp_listener(output_queue)
     rescue => e
       @logger.warn("UDP listener died", :exception => e, :backtrace => e.backtrace)
+      @metric_errors.increment(:listener_exception)
```
Comment (ph) [inline lib/logstash/inputs/udp.rb]: I am worrying about the number of the similar name of keys in the plugins, should we define an `exception` namespace and define keys in it.

Maybe something like this:

```json
{
exceptions: {
listener: 123,
worker: 1234
}
}
```

File: lib/logstash/inputs/udp.rb (hunk 2 above)
Comment (purbon) [inline lib/logstash/inputs/udp.rb]: makes sense

File: lib/logstash/inputs/udp.rb (hunk 3 above)
Comment (purbon) [inline lib/logstash/inputs/udp.rb]: I agree on having uniform patterns for namespaces, I created https://github.com/elastic/logstash/issues/5716 to centalize discussions and reach consensus 👍

File: lib/logstash/inputs/udp.rb (hunk 1 above)
Comment (purbon) [inline lib/logstash/inputs/udp.rb]: As I shared at https://github.com/logstash-plugins/logstash-input-http/pull/58#discussion_r72979337, happy to understand what is under your mind. Can you elaborate a bit more your thoughts?

File: lib/logstash/inputs/udp.rb (hunk 3 above)
Comment (purbon) [inline lib/logstash/inputs/udp.rb]: I will apply here same pattern as in https://github.com/logstash-plugins/logstash-input-syslog/pull/28

File: lib/logstash/inputs/udp.rb (hunk 1 above)
Comment (purbon) [inline lib/logstash/inputs/udp.rb]: I've been thinking a little bit around, my thoughts:

### pros

* Code is DRY'ed.

### cons / open questions

* Add metrics attributes to codecs, as we only output plugin related metrics, I wonder if there is something else that `events` to be calculated there.
* What happens when a plugin is not using a codec? code gets duplicated then. See jdbc input case.
* What happens when we move to mills? 

This makes me think that even if the option is very nice and thoughtful, adding metrics to codecs will require a bit more thinking. Unless I'm missing something for sure.

Happy to agree on something with you PH.

**Comment by purbon** [comment]
updated the PR with https://github.com/logstash-plugins/logstash-input-udp/pull/24/commits/e4ae82fcda0da9fd814a9274f29c2a5ed6716671, only open question is https://github.com/logstash-plugins/logstash-input-udp/pull/24#discussion_r73135495 happy to reach agreement on plan for this and move forward 😄

File: lib/logstash/inputs/udp.rb (hunk 1 above)
Comment (purbon) [inline lib/logstash/inputs/udp.rb]: After taking the chance to investigate the logstash code where actually codecs are being created, see https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/config/mixin.rb#L379-L410 for more details where actually codecs are being instanciated.

The only option to implement this in a nice way would be to properly "inject" the metrics object into the codec, but the logic as of now is all coupled with validating all the parameters, see https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/config/mixin.rb#L452 for exact details where codecs are right now created. As of now, without a bit of refactoring I don't see a nice was, as nice as https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/pipeline.rb#L401-L405, to sort out the "injection".

For this I do prefer that for now, we avoid having this in the codecs. we can always workout a proper solution when things like https://github.com/elastic/logstash/issues/4765 landed and we clean up this area.

Comment (jordansissel) [inline lib/logstash/inputs/udp.rb]: If we want this done in the codec, and I agree with doing that, I think we have two options for this PR to move forward:

1. Leave as-is, and later refactor once codecs are capable of counting events
2. Remove event counts from this PR.

**Comment by jordansissel** [comment]
Once the `events` count discussion is resolved, LGTM ;)

File: lib/logstash/inputs/udp.rb (hunk 1 above)
Comment (ph) [inline lib/logstash/inputs/udp.rb]: go with 1.

**Comment by ph** [comment]
LGTM, leave the count as i.

**Comment by jsvd** [comment]
Merged in #34

**Pull request closed without merging**